hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/FlinkRowDataWriteClient.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.client;
+
+import org.apache.hudi.io.v2.HandleRecords;
+import org.apache.hudi.table.action.commit.BucketInfo;
+
+import java.util.List;
+
+/**
+ * Interface for flink write client that supports writing RowData directly into the underneath filesystem.
+ * <p>
+ * todo: add support for Insert/InsertOverwrite/InsertOverwriteTable, see HUDI-9075
+ */
+public interface FlinkRowDataWriteClient {
+
+  /**
+   * Upsert a batch of new records into Hoodie table at the supplied instantTime.
+   *
+   * @param records rowdatas needed to be written, in form of iterator
+   * @param bucketInfo the bucket info for the target bucket
+   * @param instantTime instant time to commit the data
+   * @return Collection of WriteStatus to inspect errors and counts
+   */
+  List<WriteStatus> upsert(HandleRecords records, BucketInfo bucketInfo, String instantTime);
```

Comment:
Not sure about the `HandleRecords` abstraction and why `BucketInfo bucketInfo` is required for the write.

Reply:
Discussed offline:
`HandleRecords` is introduced for getting records iterator for data block and delete block  separately, since the iterator is a one-time iterator, if delete records and upsert records are mixed in one iterator, we can not divide them without materializing the iterator into List, thus optimization like object-reuse cannot be employed. 

For api design and perf trade off, we'll remove `HandleRecords`, and using one iterator contains both upsert/delete records, and divide them in log write handle.